etc/systemd/system-generators/zfs-mount-generator.in

```diff
@@ -230,15 +313,87 @@ ExecStop=@sbindir@/zfs unload-key '${dataset}'"   > "${dest_norm}/${keyloadunit}
       "${dataset}" >/dev/kmsg
   fi
 
-  # If the mountpoint has already been created, give it precedence.
+  if [ -n "${p_systemd_wantedby}" ] && \
+      [ "${p_systemd_wantedby}" != "-" ] ; then
+    p_systemd_noauto="on"
+    if [ "${p_systemd_wantedby}" = "none" ] ; then
+      wantedby=""
+    else
+      wantedby="${p_systemd_wantedby}"
+      before="${before} ${wantedby}"
+    fi
+  fi
+
+  if [ -n "${p_systemd_requiredby}" ] && \
+      [ "${p_systemd_requiredby}" != "-" ] ; then
+    if [ "${p_systemd_requiredby}" = "none" ] ; then
+      requiredby=""
+    else
+      p_systemd_noauto="on"
+      requiredby="${p_systemd_requiredby}"
+      before="${before} ${requiredby}"
+    fi
+  fi
+
+  # For datasets with canmount=on, a dependency is created for
+  # local-fs.target by default. To avoid regressions, this dependency
+  # is reduced to "wants" rather than "requires" when nofail is not "off".
+  # **THIS MAY CHANGE**
+  # p_systemd_noauto=on disables this behavior completely.
+  if [ "${p_systemd_noauto}" != "on" ] ; then
+    if [ "${p_systemd_nofail}" = "off" ] ; then
+      requiredby="local-fs.target"
+      before="${before} local-fs.target"
+    else
+      wantedby="local-fs.target"
+      if [ "${p_systemd_nofail}" != "on" ] ; then
+        before="${before} local-fs.target"
+      fi
+    fi
+  fi
+
+  # Handle existing files:
+  # 1. We never overwrite existing files, although we may delete
+  #    files if we're sure they were created by us. (see 5.)
+  # 2. Units with canmount=on always have precedence over noauto.
+  #    This is enforced by the sort pipe in the loop around this function.
+  # 3. If no unit file exists for a noauto dataset, we create one.
+  #    Additionally, we use $noauto_files to track the unit file names
+  #    (which are the systemd-escaped mountpoints) of all (exclusively)
+  #    noauto datasets that had a file created.
+  # 4. If the file to be created is found in the tracking variable,
+  #    we do NOT create it.
+  # 5. If a file exists for a noauto dataset, we check whether the file
+  #    name is in the variable. If it is, we have multiple noauto datasets
+  #    for the same mountpoint. In such cases, we remove the file for safety.
+  #    To avoid further noauto datasets creating a file for this path again,
+  #    we leave the file name in the tracking variable.
   if [ -e "${dest_norm}/${mountfile}" ] ; then
-    printf 'zfs-mount-generator: %s already exists\n' "${mountfile}" \
-      >/dev/kmsg
+    if is_known "$mountfile" "$noauto_files" ; then
+      # if it's in $noauto_files, we must be noauto too. See 2.
+      printf 'zfs-mount-generator: removing duplicate noauto %s\n' \
+        "${mountfile}" >/dev/kmsg
+      # See 5.
+      rm "${dest_norm}/${mountfile}"
+    else
+      # don't log for noauto
+      if [  "${p_systemd_noauto}" != "on" ] ; then
```

Comment:
I don't think we should suppress this log entry if `wanted-by` or `required-by` is specified.